### ISSUE TITLE
ActionSubscriber freezes

### ISSUE THREAD

**Issue by brianbroderick** [mxenabled/lifeguard#5]
This may be related to issue #2. I have 2 subscribers, which work for a while, but one of them will stop processing Rabbit messages after a random amount of time.  Which one freezes seems to be random as well. 

My Gemfile.lock shows this:

action_subscriber (2.4.0-java)
      activesupport (>= 3.2)
      lifeguard (>= 0.0.9)
      march_hare (>= 2.7.0)
      middleware
      thor

I'm using JRuby 9.1.2.0

I haven't seen an error message related to this.

**Comment by brianbroderick** [comment]
Using lifeguard 0.2.0

**Comment by mmmries** [comment]
@brianbroderick are you using `subscribe` mode or `poll` mode?

Are the two subscribers in separate processes or are they in the same JVM process?

**Comment by brianbroderick** [comment]
Subscribe mode.  They are in the same JVM process.

**Comment by brianbroderick** [comment]
```
::ActionSubscriber.configure do |config|
  config.host     = ENV["RABBITMQ_HOST"] || "localhost"
  config.port     = ENV["RABBITMQ_PORT"] || 5672
  config.username = ENV["RABBITMQ_USERNAME"]
  config.password = ENV["RABBITMQ_PASSWORD"]
  config.publisher_confirms = true
  config.threadpool_size = ENV["RABBITMQ_THREADPOOL_SIZE"].to_i || 8
  config.times_to_pop = ENV["RABBITMQ_TIMES_TO_POP"].to_i || 8
  config.prefetch = ENV["RABBITMQ_PREFETCH"].to_i || 5

  config.error_handler = lambda do |error, _env_hash|
    Rails.logger.error("Error processing message: #{error.message}\n")
    Rails.logger.error("Backtrace: #{error.backtrace.join("\n")}")
    NewRelic::Agent.notice_error(error)
    raise error
  end
end

::ActionSubscriber.draw_routes do
  route ::TextAnalysisSubscriber, :created,
        {
          acknowledgements: true,
          exchange: "events",
          publisher: "*",
          durable: true
        }
  route ::BatchAnalysisSubscriber, :created,
        {
          acknowledgements: true,
          exchange: "events",
          publisher: "*",
          durable: true
        }
  route ::CollectionAnalysisSubscriber, :created,
        {
          acknowledgements: true,
          exchange: "events",
          publisher: "*",
          durable: true
        }
end
```

**Comment by brianbroderick** [comment]
There are actually 3, but the TextAnalysis one isn't doing anything (we haven't been sending messages to it)

**Comment by brianbroderick** [comment]
FYI, I'm upgrading to 9.1.4.0 though that probably won't affect this.

**Comment by mmmries** [comment]
Glad you'll be testing out `jruby-9.1.4.0` before me 😄  Let me know how it goes.

I'll try to upgrade some more of my subscribers to the latest lifeguard today to see if we encounter the same freezing issue.

**Comment by mmmries** [comment]
@brianbroderick I just thought of something you might want to test. Can you add a call to `Thread.abort_on_exception = true` somewhere during the loading of your application?

Recently we had to rollback `jruby-9.1.2.0` because of the [Stackoverflow bug](https://github.com/jruby/jruby/issues/4020). And it occured to me last night that `action_subscriber` doesn't set this flag so threads will silently die, but setting this should crash your whole program when a thread dies so you can see what caused those threads to die. In our case the `Stackoverflow` bug would have crashed our whole subscriber process and made it easier to debug.

Not sure if I want to recommend setting this in production, but I think it could be helpful in debugging.

**Comment by brianbroderick** [comment]
Yeah, I'll give that a try.  Thanks.  Also, FYI, JRuby 9.1.4.0 had a bug that caused my app to die in CircleCI.  They released 9.1.5.0 today, which fixed at least that issue.

**Comment by mmmries** [comment]
@brianbroderick another thing you might want to try. Lifeguard `0.2.1` was released today because we found a problem with the `0.2.0` release.

It solved a pretty major freezing issue for us.
